File: project-euler/0007/0007.js

var primes = [2];
var number = 3;

while (primes.length != 10001) {
    if(isPrime(number)) {
        primes.push(number);
    }
    number += 2;
}

console.log(primes.pop());

function isPrime(number) {
    for(var i = 0; i < primes.length; i++) {
        var prime = primes[i];
        if(number % prime === 0) {
            return false;
        }
    }
    return true;
}

File: project-euler/0007/0007.pl
#!/usr/bin/env perl

sub is_prime {
    my($num) = @_;

    for(@primes) {
        if($num % $_ == 0) {
            return 0;
        }
    }
    return 1;
}

# https://en.wikipedia.org/wiki/Prime_number_theorem
$limit = 10001;
$temp_limit = $limit * log($limit) * 2;

@primes = &sieve($temp_limit);

print "@primes[$limit - 1]\n";

sub sieve {
    my($max) = @_;
    my @composite = ();
    my @prime = ();
    my $num = 2;

    while($num < $max) {
        if(@composite[$num]) {
            $num += 1;
            next;
        }
        push(@prime, $num);

        $next_num = $num + $num;
        while($next_num < $max) {
            @composite[$next_num] = 1;
            $next_num += $num;
        }
        $num += 1;
    }
    return @prime;
}
